src/audio/pipewire.c

```diff
@@ -0,0 +1,322 @@
+    uint64_t dummy; // used to read from the event file descripter. We don't use this for anything, at the moment, it's always 42
+    // if the stream has been deactivated because of a pipewire_stop call, we activate it
+    pw_thread_loop_lock(state->loop);
+    if (pw_stream_get_state(state->stream, NULL) == PW_STREAM_STATE_PAUSED)
+    {
+        pw_stream_set_active(state->stream, true);
+    }
+    pw_thread_loop_unlock(state->loop);
+    track_buffer_size = track.num_samples;
+    // as long as we have samples in the buffer sent by speech dispatcher, we loop, blocking spd from calling us till the audio system played what we loaded in it
+    while (track_buffer_size > 0)
+    {
+        // drainage spinlock!
+        // we wait till fill_quantity is not empty after this thread can continue, aka after on_process updated that ringbuffer index with what it managed to play, so that we have some room in the buffer to fill from what's remaining of what was given to us by speech dispatcher
+        while (true)
+        {
+            fill_quantity = spa_ringbuffer_get_write_index(&state->rb, &write_index);
+            // see if there's some room in the buffer, if so, stop spinning
+            room_left_in_buffer = SAMPLE_BUFFER_SIZE - fill_quantity;
+            if (room_left_in_buffer > 0)
+                // then we have room in the buffer, so stop spinning
+                break;
+            // otherwise, to not starve a cpu core of resources, we suspend our spinning until something, in this case on_process, wakes us up. In an embedded system without these primitives, spinning endlessly like that would be the only option. We could do the same here, but for efficiency reasons alone, we don't
+            pw_thread_loop_lock(state->loop);
+            struct pw_loop *program_loop = pw_thread_loop_get_loop(state->loop);
+            pw_thread_loop_unlock(state->loop);
+            spa_system_eventfd_read(program_loop->system, state->eventfd_number, &dummy);
+        }
+        // we write the amount of samples we can at this time without overflowing the buffer, to the memory area represented by the amount of room that's free after the last on_process call, to then be enqueued by pipewire
+        //  we assume speech dispatcher gives us the correct number of bytes for the format it chose, enough for this chunk. If that's not the case, there's not much we could do besides reading uninitialized memory or an incomplete chunk, unfortunately
+        // if we are about to overflow the buffer, better sharply clamp it than cause a buffer overflow vulnerability
+        room_left_in_buffer = SPA_MIN(SAMPLE_BUFFER_SIZE, room_left_in_buffer);
+        spa_ringbuffer_write_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, write_index % SAMPLE_BUFFER_SIZE, track.samples, room_left_in_buffer);
+        // subtract from the total buffer what we were already able to write, in order to indicate that the buffer  has more room
```

Comment:
no, it's not to indicate that the buffer has more room, but rather that we have progressed in pushing the audio to `sample_buffer`